src/os/bluestore/BlueFS.cc

```diff
@@ -833,20 +834,36 @@ int BlueFS::_check_new_allocations(const bluefs_fnode_t& fnode,
 }
 
 int BlueFS::_adjust_granularity(
-  __u8 id, uint64_t *offset, uint64_t *length, const char *op)
+  __u8 id, uint64_t *offset, uint64_t *length, bool alloc)
 {
+  const char *op = alloc ? "op_alloc_add" : "op_alloc_rm";
   auto oldo = *offset;
   auto oldl = *length;
   if (*offset & (alloc_size[id] - 1)) {
     *offset &= ~(alloc_size[id] - 1);
     *offset += alloc_size[id];
     if (*length > *offset - oldo) {
+      if (alloc) {
+	block_unused_too_granular[id].insert(oldo, *offset - oldo);
+      } else {
+	block_unused_too_granular[id].erase(oldo, *offset - oldo);
+      }
       *length -= (*offset - oldo);
     } else {
+      if (alloc) {
+	block_unused_too_granular[id].insert(oldo, *length);
+      } else {
+	block_unused_too_granular[id].erase(oldo, *length);
+      }
       *length = 0;
     }
   }
-  *length &= ~(alloc_size[id] - 1);
+  if (*length & (alloc_size[id] - 1)) {
```

Comment:
Shouldn't we process !alloc case and remove from the list here as well?

Reply:
yep!